service/http/http.go

```diff
@@ -299,3 +348,78 @@ type Component interface {
 	// will receive a request to just `/metrics`.
 	Handler() http.Handler
 }
+
+// lazyListener is a [net.Listener] which lazily initializes the underlying
+// listener.
+type lazyListener struct {
+	mut    sync.RWMutex
+	inner  net.Listener
+	closed bool
+}
+
+var _ net.Listener = (*lazyListener)(nil)
+
+// SetInner updates the inner listener. It is safe to call SetInner multiple
+// times. SetInner panics if given a nil argument.
+//
+// SetInner returns an error if called after the listener is closed.
+func (lis *lazyListener) SetInner(inner net.Listener) error {
+	if inner == nil {
+		panic("Unexpected nil listener passed to SetInner")
+	}
+
+	lis.mut.Lock()
+	defer lis.mut.Unlock()
+
+	if lis.closed {
+		return net.ErrClosed
+	}
+
+	lis.inner = inner
+	return nil
+}
+
+func (lis *lazyListener) Accept() (net.Conn, error) {
+	// The read lock is held as briefly as possible since Accept is a blocking
+	// call and may hold the read lock longer than we want it to.
+	lis.mut.RLock()
+	var (
+		inner  = lis.inner
+		closed = lis.closed
+	)
+	lis.mut.RUnlock()
+
+	if closed || inner == nil {
+		return nil, net.ErrClosed
+	}
+	return inner.Accept()
+}
+
+func (lis *lazyListener) Close() error {
+	lis.mut.Lock()
+	defer lis.mut.Unlock()
+
+	if lis.closed {
+		return net.ErrClosed
+	}
+
+	lis.closed = true
+	return lis.inner.Close()
+}
+
+func (lis *lazyListener) Addr() net.Addr {
+	lis.mut.RLock()
+	defer lis.mut.RUnlock()
+
+	if lis.inner == nil {
+		// TODO(rfratto): it's not sure if this will cause problems. If this is an
+		// issue, we can do one of two things to address this:
+		//
+		// 1. Return a fake address.
```

Comment:
I'm just not sure this code ever gets called, which is why I'm leaving it as a TODO. Option 2 would take more time to implement since it's not simple, so I'd prefer to leave this as-is and we can revisit if it turns out we do hit panics.

Reply:
I see. Since its a new feature, it's probably fine as-is. Would be good idea to start using it right away in some test deployments.